sw/device/tests/crypto/kdf_hmac_ctr_functest.c

```diff
@@ -0,0 +1,535 @@
+  otcrypto_blinded_key_t kdk = {
+      .config = kdk_config,
+      .keyblob = kdk_keyblob,
+      .keyblob_length = sizeof(kdk_keyblob),
+  };
+  kdk.checksum = integrity_blinded_checksum(&kdk);
+
+  // Construct a blinded key struct for the output keying material. The key mode
+  // here doesn't really matter, it just needs to be some symmetric key.
+  otcrypto_key_config_t km_config = {
+      .version = kOtcryptoLibVersion1,
+      .key_mode = test->km_mode,
+      .key_length = test->km_bytelen,
+      .hw_backed = kHardenedBoolFalse,
+      .exportable = kHardenedBoolFalse,
+      .security_level = kOtcryptoKeySecurityLevelLow,
+  };
+  uint32_t km_keyblob[keyblob_num_words(km_config)];
+  otcrypto_blinded_key_t km = {
+      .config = km_config,
+      .keyblob = km_keyblob,
+      .keyblob_length = sizeof(km_keyblob),
+  };
+
+  // Construct a buffer for the context.
+  otcrypto_const_byte_buf_t context = {
+      .data = test->kdf_context,
+      .len = test->kdf_context_bytelen,
+  };
+
+  // Construct a buffer for the label.
+  otcrypto_const_byte_buf_t label = {
+      .data = test->kdf_label,
+      .len = test->kdf_label_bytelen,
+  };
+
+  // Run the KDF specified by the key mode.
+  switch (test->key_mode) {
+    case kOtcryptoKeyModeHmacSha256:
+    case kOtcryptoKeyModeHmacSha384:
+    case kOtcryptoKeyModeHmacSha512:
+      TRY(otcrypto_kdf_hmac_ctr(kdk, label, context, km.config.key_length,
+                                &km));
+      break;
+    default:
+      LOG_INFO("Should never end up here.");
+      return INVALID_ARGUMENT();
+  }
+
+  LOG_INFO("KDF operation completed.");
+
+  // Unmask the output key value and compare to the expected value.
+  uint32_t *km_share0;
+  uint32_t *km_share1;
+
+  TRY(keyblob_to_shares(&km, &km_share0, &km_share1));
+  uint32_t unmasked_km[keyblob_share_num_words(km_config)];
+  for (size_t i = 0; i < ARRAYSIZE(unmasked_km); i++) {
+    unmasked_km[i] = km_share0[i] ^ km_share1[i];
+  }
+
+  TRY_CHECK_ARRAYS_EQ((unsigned char *)unmasked_km,
+                      (unsigned char *)test->keying_material, test->km_bytelen);
+  return OK_STATUS();
+}
+
+/**
+ * Test case 1:
+ *
+ * Basic test case with HMAC SHA256
+ *
+ * KDF Mode = HMAC Counter
+ * KDK      = 0x0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b0b0b0b0b0b0b0b
+ *              0b0b0b0b0b0b0b0b (128 octets)
+ * context  = 0x0102030405060708090a0b0c (12 octets)
+ * label    = 0xf0f1f2f3f4f5f6f7f8f9 (10 octets)
+ * L        = 256
+ *
+ * KM       = 0x6caa24e6f52b58092b89cc38807035
+ *              1d91be9196aa931aff8694f31e6dc2
+ *              9de8 (32 octets)
+ */
+static status_t func_test1(void) {
```

Comment:
nit: I'd give these more descriptive names. Did you get the tests from a specific source? If so and they have numbers from that source then naming them that way might make sense e.g. `nist_test1`. Otherwise, I'd try something descriptive like `sha256_short_tag_test`. That way if something starts failing, it's easy to see in the error message if e.g. all the sha256 tests are failing or all the tests with short tags are failing.

Reply:
How exactly did you create the test values? Did you run them through another library?

Reply:
Regarding the functional tests, since there are no test vectors in the specification, as in the case of HKDF, I used sw/device/tests/crypto/hkdf_functest.c as a template and generated keying material with a python script (pycryptodome library).